Feat Proposal: add progressbar to output

dimtrovich opened the pull request:
it would be nice to have a progress bar in the console. At times, a command can execute several tasks simultaneously (for example creation of controllers, models, views, etc.). It would therefore be practical for the person opposite to know the current level of progress of the activity.

adhocore commented:
thank you very much. checking soon. this is definitely a missing and good to have stuff. 👍

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -0,0 +1,300 @@
+     */
+    public function forceRedraw(bool $force = true): self
+    {
+        $this->forceRedraw = !!$force;
```
`!!` may not be required

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -0,0 +1,300 @@
+            $items = iterator_to_array($items);
+        }
+
+        $total = count($items);
```
```suggestion
        if (0 === $total = count($items)) {
            return;
        }
```

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -0,0 +1,300 @@
+     */
+    public function forceRedraw(bool $force = true): self
+    {
+        $this->forceRedraw = !!$force;
```
```suggestion
        $this->forceRedraw = $force;
```

dimtrovich replied:
well seen. I don't even know why I wrote it like that 😂

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -0,0 +1,300 @@
+    {
+        $percentage = $current / $this->total;
+        $bar_length = round($this->getBarStrLen() * $percentage);
+
```
```suggestion
```

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -0,0 +1,300 @@
+
+        if ($label) {
+            $label = $this->labelFormatted($label);
+        // If this line doesn't have a label, but we've had one before,
```
```suggestion
            // If this line doesn't have a label, but we've had one before,
            // then ensure the label line is cleared
```

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -0,0 +1,300 @@
+
+    /**
+     * Get the progress bar string, basically:
+     * =============>             50% label
```
pretty nice. it would be even better if we can choose progressbar char from a set eg `=`, `■`, `▧`, `▨`, `▦`, `▩`  etc with default `=` :) 
which seems as simple as swapping the `options['loader']`

dimtrovich replied:
that's the idea of the options property. it is planned to put a setter which will modify the pointer, the loader and the color of the progress bar.
we can do something like 
`$progress->option('loader', '*');` or 
`$progress->option([
    'loader' => '*',
    'pointer' => '>>'
]);`

adhocore commented on src/Input/Command.php:
```diff
@@ -382,4 +385,20 @@ protected function io(): Interactor
+    protected function progress(int $total = null): ProgressBar
+    {
+        if ($this->_progressBar === null) {
+            $this->_progressBar = new ProgressBar(null, $this->writer());
```
is same progressbar reusable again n again without resetting?

dimtrovich replied:
oh no... i was too focused on minimal resource usage but this way we'll have the same instance of the progressbar every time. thanks very much.

Nevertheless I do not see why someone would venture to call this method several times in the same command

adhocore replied:
yea that may not be too normal but not impossible either. maybe in rare cases someone would like to have multiple phases and each phase with different steps that would show progresses. 
we can revisit on this later though.

adhocore reviewed (commented):
thanks again, looks good :100:. liked how the code quality is aligned to the project. 
only missing is unit test :D

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -166,7 +166,9 @@ public function each($items, callable $callback = null)
         }
 
-        $total = count($items);
+        if (0 === $total = count($items)) {
```
ooh oops, i think i didn't know how to format code suggestion of github.
the idea was to combine assignment and checks in oneliner
```php
$total = count($items) 
if (!$total) {
    return;
}

if (0 === $total = count($items)) {
    return;
}
```

adhocore commented:
hello @dimtrovich , not sure if you lost interest on this pull request? :)

dimtrovich commented:
hello @adhocore. sorry for the absence, I had a few small occupations.

I added progress bar customization via options and removed singletton behavior from progress bar initialization in commands.

however I had difficulty writing tests. it must be said that I do not use phpunit too much but rather kahlan or pest.
It will make me happy if you give me a hand for setting up the tests, it will allow me to learn a little more 🙏🙏

adhocore commented on src/Output/ProgressBar.php:
```diff
@@ -0,0 +1,319 @@
+        $line_count = $this->hasLabelLine ? 2 : 1;
+
+        $progress_bar  = $this->cursor->up($line_count);
+        $progress_bar .= "\r";
```
this this CR char required? in unix it doesnt seem to be required

(i'm on branch [81-progress-bar](https://github.com/adhocore/php-cli/blob/81-progress-bar) trying to write tests)

dimtrovich replied:
I haven't tried with unix but with windows it is essential. when removed, the progress bar goes to a new line on each iteration

Preview => https://ibb.co/W0kMTcz
